feat: switch to condensed view based on container width

geoffrich opened the pull request:
# Alaska Airlines Pull Request

Fixes https://github.com/AlaskaAirlines/auro-flight/issues/19

## Summary:

- When an individual flightline's width drops below 414px, it switches to the condensed view seen on mobile. This breakpoint is customizable via the `breakpoint` property on the flightline.
- Use ResizeObserver to detect changes to element size. The RO is shared between all flightline instances on a page.
- Refactor: remove --layover class. This was always added, so we didn't need a separate declaration.

Dale suggested looking into the cqfill package to polyfill CSS container queries. However, I was unable to get that package to work with web components (see discussion in linked auro-flight issue). I couldn't find any other stable polyfill, options, so I opted to keep my original ResizeObserver solution instead. This should be relatively easy to swap out for container queries when they're available in browsers -- instead of setting a `condensed` class, use a media query instead and remove the observer code.

If anyone has suggestions for alternate solutions for me to investigate, let me know.

Before (675px screen width):
![image](https://user-images.githubusercontent.com/4992896/138176334-af61d13b-e57c-4893-9493-d80cbeb0be23.png)

After:
![image](https://user-images.githubusercontent.com/4992896/138176423-3bb2b45f-9651-4034-af4f-c3480fbcb6c9.png)


As a demonstration, add the following HTML to the demo and resize the screen.

```html
<style>
  .grid {
    display: grid;
    grid-template-columns: 1fr 1fr;
    gap: 2rem;
  }
</style>

<div class="grid">
  <auro-flightline id="flight1">
    <auro-flight-segment stopover iata="YAK"></auro-flight-segment>
    <auro-flight-segment stopover iata="WRG"></auro-flight-segment>
    <auro-flight-segment iata="SEA" duration="0h 40m"></auro-flight-segment>
    <auro-flight-segment iata="BOS" duration="1h 40m"></auro-flight-segment>
    <auro-flight-segment iata="DUB" duration="13h 40m"></auro-flight-segment>
  </auro-flightline>
  <auro-flightline id="flight2">
    <auro-flight-segment stopover iata="YAK"></auro-flight-segment>
    <auro-flight-segment stopover iata="WRG"></auro-flight-segment>
    <auro-flight-segment iata="SEA" duration="0h 40m"></auro-flight-segment>
    <auro-flight-segment iata="BOS" duration="1h 40m"></auro-flight-segment>
    <auro-flight-segment iata="DUB" duration="13h 40m"></auro-flight-segment>
  </auro-flightline>
</div>
```

## Type of change:

Please delete options that are not relevant.

- [x] New capability

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

blackfalcon commented:
Given this draft PR, how do you feel about working with this polyfill that is based on the emerging spec? 

https://github.com/AlaskaAirlines/auro-flight/issues/19#issuecomment-932537409

geoffrich commented on src/auro-flightline.js:
```diff
@@ -10,6 +10,9 @@ import { classMap } from "lit-html/directives/class-map";
 import styleCss from "./style-flightline-css.js";
+import { observeResize, unobserve } from './observer';
+
+const SMALL_BREAKPOINT = 660;
```
Latest commit names this with camel case.

geoffrich commented on src/observer.js:
```diff
@@ -0,0 +1,54 @@
+}
+
+// this observer is shared between all auro-flightline instances on a page
+const ro = new ResizeObserver(handleResize);
```
Added the above snippet and this seems to work now.

geoffrich commented on src/auro-flightline.js:
```diff
@@ -36,6 +43,19 @@ class AuroFlightline extends LitElement {
+  firstUpdated() {
+    // if the container's width drops below the small breakpoint, force the condensed view
+    const setCondensed = (val) => {
+      this.condensed = val < SMALL_BREAKPOINT;
```
I refactored the code to be mobile-first, swapping `condensed` for `expanded`. Styles are simpler now!

geoffrich commented on src/auro-flight-segment.js:
```diff
@@ -49,16 +49,13 @@ class AuroFlightSegment extends LitElement {
       <div class="layout">
         <slot></slot>
-        <div class="${classMap(legClasses)}"></div>
+        <div class="leg layout ${classMap(legClasses)}"></div>
```
Reverted this change in latest commit.

geoffrich commented on src/auro-flightline.js:
```diff
@@ -44,11 +64,11 @@ class AuroFlightline extends LitElement {
 
     return html`
-      <div class="${classMap(classes)}">
+      <div class="slot-container ${classMap(classes)}">
```
Reverted this change in latest commit.

geoffrich commented on src/style-flightline.scss:
```diff
@@ -43,6 +44,14 @@
+.condensed.multiple auro-flight-segment {
+  display: block;
+}
+
```
Do my latest changes address your concerns here? Now the styles mirror the old code, just replacing media queries with `.expanded`.

geoffrich commented:
@blackfalcon finally got back to this. I think I addressed all your comments -- please take another look. Going mobile-first made things a lot cleaner.

Also, I think you linked the wrong issue to this PR -- this addresses https://github.com/AlaskaAirlines/auro-flight/issues/19. The canceled flightline will be addressed separately.

blackfalcon commented on src/style-flightline.scss:
```diff
@@ -33,28 +37,23 @@
-  @include auro_breakpoint--sm {
-    display: block;
-  }
+.multiple.expanded ::slotted(*) {
```
```suggestion
.multiple.switch ::slotted(*) {
```

blackfalcon commented on test/auro-flightline.test.js:
```diff
@@ -54,7 +54,40 @@ describe('auro-flightline', () => {
+    await expect(el).to.be.true;
+  });
+
+  it('shows expanded view depending on width', async () => {
```
```suggestion
  it('shows switched view depending on width', async () => {
```

blackfalcon commented on src/auro-flightline.js:
```diff
@@ -10,23 +10,33 @@ import { classMap } from "lit-html/directives/class-map";
 import styleCss from "./style-flightline-css.js";
+import { observeResize, unobserve } from './observer';
+
+const defaultBreakpoint = 414;
```
```suggestion
const defaultContainer= 414;
```

blackfalcon commented on src/observer.js:
```diff
@@ -0,0 +1,54 @@
+/** @type {Map<Node, resizeCallback>} */
```
Something I have been thinking about is making utilities like these Class declarations so that we can ship them in a utility package and then users can extend the features if needed within the individual setup. 

Then each project can have these installed via NPM and we can have easier control over them when updated. 

Thoughts?

geoffrich replied:
I would prefer to move these into a common package when it becomes necessary for another component to use this functionality. Otherwise it's just another package + repo to maintain if we don't have the immediate need to share this code.

geoffrich commented on src/auro-flightline.js:
```diff
@@ -10,23 +10,33 @@ import { classMap } from "lit-html/directives/class-map";
-      canceled:   { type: Boolean }
+      canceled:    { type: Boolean },
+      expanded:    { type: Boolean },
+      breakpoint:  { type: Number }
```
Is the reasoning behind making a string just so the user better understands it's a px value? I would have concerns that if we expose it that way then they might think they can do other units as well like 50rem, 50%, etc, which would be hard to validate. Having it be a number is just simpler and we can clearly explain in the API documentation that it represents a pixel value.

geoffrich commented on src/auro-flightline.js:
```diff
@@ -10,23 +10,33 @@ import { classMap } from "lit-html/directives/class-map";
+    this.breakpoint = defaultBreakpoint;
+
+    /** @private */
+    this.expanded = false;
```
I was thinking of it as "expanded view", i.e. showing all the stops. "switch" feels just as confusing to me.

What about `showAllStops`? Since that's what the component is doing.

Naming things is hard.

geoffrich replied:
Renamed to "showAllStops". I'm going to resolve the other comments around this name change, but leave this one open for further discussion.

geoffrich commented:
Refactored according to suggested name changes. See my comments above. Also updated the tests to ignore the "ResizeObserver limit exceeded" error, which only happened in the tests. Previously, I had wrapped the call in `requestAnimationFrame` to fix this, but this was inexplicable causing several tests to be skipped. I removed the `requestAnimationFrame` and told Mocha to treat the error as a warning instead.

@blackfalcon let me know what you think of the name changes (though I know you're off for the holidays soon, so no rush)

blackfalcon commented:
I am approving this work given the scope is directly tied to this element's use. Abstracting this into an individual utility will be discussed in the future with a different proposal of work. 

Given the current maturity of the `@content` selector and its current availability [behind a chrome flag](https://css-tricks.com/say-hello-to-css-container-queries/), it's safe to say that by the time we get to the work needed to make this a shareable utility the official spec will be in the wild.
